### ISSUE TITLE
[rush] Internal Error "Cannot find installed dependency" with PNPM 7

### ISSUE THREAD

**Issue by jeremymeng** [microsoft/rushstack#3652]
https://github.com/microsoft/rushstack/issues/2542#issuecomment-1133416811 is closed as completed so logging a new issue since  I am still hitting it with PNPM v7, and currently we don't set `useWorkspace=true` (for a couple of reasons)

## Summary

when running `rush update`, rush seems looking for local projects in the wrong directories.

## Repro steps
1. clone `https://github.com/jeremymeng/azure-sdk-for-js/tree/engsys/bump-rush-pnpm-version`
2. cd azure-sdk-for-js
3. rush update --full

  **Actual result:** 

>Linking local projects

>LINKING: @azure-rest/agrifood-farming

>ERROR: Internal Error: Cannot find installed dependency "@azure-tools/test-recorder" in
"/workspaces/js-main/common/temp/node_modules/.pnpm/local++workspaces+js-main+common+temp+projects+agrifood-farming.tgz/node_modules"
You have encountered a software defect. Please consider reporting the issue to the maintainers of this application.

## Details
Looks that the local folder created by PNPM is under `file+projects+agrifood-farming.tgz` directory in PNPM v7, but not sure why rush is looking under `local++workspaces+js-main+common+temp+projects+agrifood-farming.tgz`

```
$ ll /workspaces/js-main/common/temp/node_modules/.pnpm/*agri*
/workspaces/js-main/common/temp/node_modules/.pnpm/file+projects+agrifood-farming.tgz:
total 80
drwxrwxrwx+    3 codespace codespace  4096 Sep 26 19:25 ./
drwxrwxrwx+ 1468 codespace codespace 69632 Sep 26 19:25 ../
drwxrwxrwx+    7 codespace codespace  4096 Sep 26 19:25 node_modules/

```

### issues trying `useWorkspaces=true`
* We would need to use `decoupledLocalDependencies` for most of our packages. currently we have over 300 of them.
* Some dependencies in package.json files are updated to have `workspace:`, which adds code churn

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.78.1 |
| `rushVersion` from rush.json? | 5.63.0 |
| `useWorkspaces` from rush.json? | not set |
| Operating system? | Linux, probably same on all systems |
| Would you consider contributing a PR? | No |
| Node.js version (`node -v`)? | 16.13.1 |

**Comment by iclanton** [comment]
I'll clone your repo and take a look, but your issues with `useWorkspaces=true` are surprising to me.

>We would need to use decoupledLocalDependencies for most of our packages. currently we have over 300 of them.

Why do so many of your packages refer to previously-published versions of other packages instead of the workspace versions?

>Some dependencies in package.json files are updated to have workspace:, which adds code churn

The cool thing about the `workspace:` version protocol is that it *reduces* code churn long-term, since it

**Comment by jeremymeng** [comment]
> Why do so many of your packages refer to previously-published versions of other packages instead of the workspace versions?

Since we have the semantic version range, say `"depB": "^1.0,0" in package A, usually we don't update package A's depB version to "^1.1.0" unless package A needs to use the new feature that is introduced in depB v1.1.0.  It should work with either published version or local workspace version.  We can release package A without having to release depB v1.1.0 first.

> The cool thing about the `workspace:` version protocol is that it _reduces_ code churn long-term, since version bumps and publishes will only bump the `version` field in packages that are published, but don't need to change the version selectors in downstream packages.

I will need to learn more about "workspace:" version protocol.  I had thought that it's PNPM specific and wasn't sure if we should have it in our published packages.

**Comment by iclanton** [comment]
>I will need to learn more about "workspace:" version protocol. I had thought that it's PNPM specific and wasn't sure if we should have it in our published packages.

It gets replaced with the dependency's version during publish.

**Comment by xirzec** [comment]
We also have cases where we want to migrate our monorepo incrementally, e.g. we major version a shared component and adopt it incrementally across dependants.

Since we're still prevented from updating pnpm and hitting some component governance issues, is there any suggested workaround we can try to unblock ourselves?

**Comment by jeremymeng** [comment]
@iclanton It looks that rush has some logic to handle pnpm v7 local packages https://github.com/microsoft/rushstack/blob/%40microsoft/rush_v5.88.0/libraries/rush-lib/src/logic/pnpm/PnpmLinkManager.ts#L276-L304

The logic may not apply for pnpm v7.  I haven't found the pnpm commit that changed the folder name from `local+...` to `file+...`. However, looking at their changelog, at least the hashing part has been changed too:

> Use a base32 hash instead of a hex to encode too long dependency paths inside node_modules/.pnpm https://github.com/pnpm/pnpm/pull/4552.

I patched rush code locally to handle pnpm version 7.  It seems working fine in our repo.  However, I wonder for such dramastic change, how comes no one else encountered it?  Has everyone moved to use workspaces so they don't have this issue?

**Comment by iclanton** [comment]
@jeremymeng I think most people are using workspaces.

Would you like to contribute your fix in a PR?

**Comment by jeremymeng** [comment]
> @jeremymeng I think most people are using workspaces.
> 
> Would you like to contribute your fix in a PR?

I can give it a shot.  I don't know what proper testing I need though.

**Comment by jeremymeng** [comment]
Fixed in PR #3908
